internal/events/tokens_approved.go

```diff
@@ -54,61 +56,71 @@ func (em *eventManager) loadApprovalOperation(ctx context.Context, tx *fftypes.U
 }
 
 func (em *eventManager) persistTokenApproval(ctx context.Context, approval *tokens.TokenApproval) (valid bool, err error) {
-	pool, err := em.database.GetTokenPoolByProtocolID(ctx, approval.Connector, approval.PoolProtocolID)
+	// Check that this is from a known pool
+	// TODO: should cache this lookup for efficiency
+	pool, err := em.database.GetTokenPoolByLocator(ctx, approval.Connector, approval.PoolLocator)
 	if err != nil {
 		return false, err
 	}
 	if pool == nil {
-		log.L(ctx).Infof("Token approval received for unknown pool '%s' - ignoring: %s", approval.PoolProtocolID, approval.Event.ProtocolID)
+		log.L(ctx).Infof("Token approval received for unknown pool '%s' - ignoring: %s", approval.PoolLocator, approval.Event.ProtocolID)
 		return false, nil
 	}
 	approval.Namespace = pool.Namespace
 	approval.Pool = pool.ID
 
+	// Check that approval has not already been recorded
+	if existing, err := em.database.GetTokenApprovalByProtocolID(ctx, approval.Pool, approval.ProtocolID); err != nil {
+		return false, err
+	} else if existing != nil {
+		log.L(ctx).Warnf("Token approval '%s' has already been recorded - ignoring", approval.ProtocolID)
+		return false, nil
+	}
+
 	if approval.TX.ID != nil {
 		if err := em.loadApprovalOperation(ctx, approval.TX.ID, &approval.TokenApproval); err != nil {
 			return false, err
 		}
-
 		if valid, err := em.txHelper.PersistTransaction(ctx, approval.Namespace, approval.TX.ID, approval.TX.Type, approval.Event.BlockchainTXID); err != nil || !valid {
 			return valid, err
 		}
 
-		existing, err := em.database.GetTokenApproval(ctx, approval.Connector, approval.ProtocolID, pool.ID)
+		existingLocal, err := em.database.GetTokenApprovalByID(ctx, approval.LocalID)
 		if err != nil {
 			return false, err
 		}
-
-		// If there's not an existing approval, look for approvals with a matching local ID
-		if existing == nil {
-			existingLocal, err := em.database.GetTokenApprovalByID(ctx, approval.LocalID)
-			if err != nil {
-				return false, err
-			}
-
-			if existingLocal != nil {
-				approval.LocalID = fftypes.NewUUID()
-			}
-		} else {
-			log.L(ctx).Infof("Updating existing approval with local ID: %s", existing.LocalID)
-			approval.LocalID = existing.LocalID
+		if existingLocal != nil {
+			approval.LocalID = fftypes.NewUUID()
```

Comment:
A comment here would be worthwhile. My understanding is:

```
// The approval was initiated by FireFly and one UUID was allocated as a `localId` for the approval when it was
// initiated. This allows the `confirm=true` code to know when the first approval record gets inserted into the
// database that was correlated to that approval. However, approvals can result in many approval records
// depending on the on-chain mapping of the scope of an approval. So the connector might generate multiple
// notifications (with different `protocolId`), but all will reach this code with the same `localId`.
// So we need to generate a new one (for all apart from the first)
```

Reply:
Yep, I'll add some notes

Reply:
Added a lot of notes and also cleaned up the method further to (hopefully) make it clearer to follow.